src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+
+    let mut attributes = BTreeMap::new();
+    let custom_attributes: Vec<String> = log
+        .keys()
+        .filter(|path| !RESERVED_ATTRIBUTES.contains(path.as_str()))
+        .collect();
+    for path in custom_attributes {
+        if let Some(value) = log.remove(&path) {
+            attributes.insert(path, value);
+        }
+    }
+    log.insert("attributes", attributes);
+
+    let bytes = serde_json::to_vec(&log)
+        .map(|mut b| {
+            b.push(b'\n');
+            b
+        })
+        .expect("Failed to encode event as json, this is a bug!");
+
+    Some(EncodedEvent {
+        item: PartitionInnerBuffer::new(bytes, key.into()),
+        finalizers: log.metadata_mut().take_finalizers(),
+    })
+}
+
+/// Generates a unique ID compatible with DD
+fn generate_log_id() -> String {
+    let mut id = BytesMut::with_capacity(18);
+    // timestamp in millis - 6 bytes
+    let now = Utc::now();
+    id.put_int(now.timestamp_millis(), 6);
+    // namespace + version, both 0 - 2 bytes
+    id.put_i16(0);
+
+    // host-based number - 5 bytes
+    lazy_static! {
+        static ref HOST_NUMBER: Vec<u8> = host_number();
+    }
+    id.put_slice(&HOST_NUMBER[..]);
+
+    // counter - 5 bytes
+    id.put_u8(0);
+    // one padding byte
+    // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+    lazy_static! {
+        static ref COUNTER: CounterU32 = CounterU32::new(0);
+    }
+    id.put_u32(COUNTER.inc()); // 4 bytes
+
+    base64::encode(id.freeze())
+}
+
+/// Returns 5 first bytes of the MAC address XOR-ed with random bytes
+fn host_number() -> Vec<u8> {
+    extern crate mac_address;
+    use mac_address::get_mac_address;
+    use rand::Rng;
+
+    let mac = get_mac_address().unwrap().unwrap().bytes();
+    let mut rng = rand::thread_rng();
+    let mut rand_bytes = rng.gen::<[u8; 5]>();
+
+    rand_bytes
+        .iter_mut()
+        .zip(mac[..5].iter()) // take first 5 bytes
+        .map(|(x, y)| *x ^ *y) // apply XOR
+        .collect::<Vec<_>>()
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "datadog_archives")]
+impl SinkConfig for DatadogArchivesSinkConfig {
+    async fn build(
+        &self,
+        cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink_and_healthcheck = self.new(cx)?;
+        Ok(sink_and_healthcheck)
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "datadog_archives"
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use chrono::DateTime;
+
+    #[test]
+    fn generate_config() {
+        crate::test_util::test_generate_config::<DatadogArchivesSinkConfig>();
+    }
+
+    #[test]
+    fn encodes_event() {
+        let mut log = Event::from("test message");
+        log.as_mut_log().insert("service", "test-service");
+        log.as_mut_log().insert("not_a_reserved_attribute", "value");
```

Comment:
Prefer to create this as a log event directly:
```suggestion
        let mut log = LogEvent::from("test message");
        log.insert("service", "test-service");
        log.insert("not_a_reserved_attribute", "value");
```
Then, below, when it needs to be an `Event`, you can use `log.into()`.